### ISSUE TITLE
Synchronized usage of Concurrent Multi Hash Map

### ISSUE THREAD

**Issue by ramadasa** [google/guava#5337]
While looking for a Concurrent Multi Hash Map, I found `Multimaps.synchronizedMultimap()` wrapper which seems to be what I'm looking for.
However, the documentation gives an example which I'm not picturing a situation when this might be useful:
https://github.com/google/guava/blob/4709fe4d50cf5249b9e5e40fb8917c610831c317/guava/src/com/google/common/collect/Multimaps.java#L566-L583

My take is that `put` operations will still be outside this synchronized block. Am I getting it wrong?

I went to look for valid usages in the test suite, but [only found one](https://github.com/google/guava/blob/0cd4e9faa1360da4a343f84cb275d6eda0c5e732/android/guava-tests/test/com/google/common/collect/MultimapsTest.java#L911-L922) that didn't help me much understanding it.

My use case is that I don't really care for non-deterministic behavior as long as it does not result in a fail operation (e.g., throw an Exception). Even if this might not be an issue, I am still curious about a scenario where this should be employed.

**Comment by ramadasa** [comment]
To add a bit more context as my initial phrasing is not fully explicit.

[All methods of `SynchronizedMultimap` are synchronized](https://github.com/google/guava/blob/064fac35d71231aba35062d1965983ecd36b6873/guava/src/com/google/common/collect/Synchronized.java#L573-L757) with a [non-accessible mutex](https://github.com/google/guava/blob/064fac35d71231aba35062d1965983ecd36b6873/guava/src/com/google/common/collect/Synchronized.java#L71). This implementation leads to the user to expect no explicit synchronization as the library already takes care of it. Hence, there is no need for the `put` or even `get` operations to be externally synchronized by the user. In fact, although not explicit (I hope I am not missing it anywhere), the [returned collection](https://github.com/google/guava/blob/064fac35d71231aba35062d1965983ecd36b6873/guava/src/com/google/common/collect/Synchronized.java#L629) of `get` is [synchronized](https://github.com/google/guava/blob/064fac35d71231aba35062d1965983ecd36b6873/guava/src/com/google/common/collect/Synchronized.java#L906-L918) which makes it even easier for the user to deal with concurrency issues.

Now, we know that we can make use of the API without incurring into explicit locking mechanisms (the library will take care of it for us). So, why does the documentation raises that concern?

**Comment by ramadasa** [comment]
While providing further context I ended up founding the solution 🤦 I opted to still posting it to share my reasoning hoping it can still be helpful to others. I'm also going to close this issue, but do correct me if I'm wrong.

While looking around I never questioned what happened when the [constructor passed a `null` mutex](https://github.com/google/guava/blob/4709fe4d50cf5249b9e5e40fb8917c610831c317/guava/src/com/google/common/collect/Multimaps.java#L594). If you follow the method calls you'll end up in the `SynchronizedObject` constructor which has the answer:
https://github.com/google/guava/blob/064fac35d71231aba35062d1965983ecd36b6873/guava/src/com/google/common/collect/Synchronized.java#L75
The object itself is the mutex. So, when the documentation advises to lock on `multimap`, the user is in fact using that mutex that I previously mentioned not having access. Now, the documentation makes sense. 🙂 
> Failure to follow this advice may result in non-deterministic behavior.

☝️ This is not referring to possible exceptions. Instead it refers to `put` operations being able to modify the view. The later was clear to me, but not the former. All good now.

So, for my use case I can safely ignore the advice the documentation shows.

Just thinking that I went on even looking for the commit message that added this documentation makes me chuckle. I was missing something so simple! I had some fun throughout the process and even learnt some stuff. Next time I'll know better 😄
